core/lb.go

```diff
@@ -0,0 +1,202 @@
+package core
+
+import (
+	"errors"
+	"math"
+	"strings"
+	"sync"
+
+	"github.com/golang/glog"
+
+	"github.com/livepeer/go-livepeer/common"
+	"github.com/livepeer/lpms/ffmpeg"
+)
+
+var ErrTranscoderBusy = errors.New("TranscoderBusy")
+
+type TranscoderSession interface {
+	Transcoder
+	Stop()
+}
+
+type newTranscoderFn func(device string, workDir string) TranscoderSession
+
+type LoadBalancingTranscoder struct {
+	transcoders []string // Slice of device IDs
+	workDir     string
+	newT        newTranscoderFn
+
+	// The following fields need to be protected by the mutex `mu`
+	mu       *sync.RWMutex
+	load     map[string]int
+	sessions map[string]*transcoderSession
+	idx      int // Ensures a non-tapered work distribution
+}
+
+func NewLoadBalancingTranscoder(devices string, workDir string, newTranscoderFn newTranscoderFn) Transcoder {
+	d := strings.Split(devices, ",")
+	return &LoadBalancingTranscoder{
+		transcoders: d,
+		workDir:     workDir,
+		newT:        newTranscoderFn,
+		mu:          &sync.RWMutex{},
+		load:        make(map[string]int),
+		sessions:    make(map[string]*transcoderSession),
+	}
+}
+
+func (lb *LoadBalancingTranscoder) Transcode(job string, fname string, profiles []ffmpeg.VideoProfile) (*TranscodeData, error) {
+
+	lb.mu.RLock()
+	session, exists := lb.sessions[job]
+	lb.mu.RUnlock()
```

Comment:
This will allow creation of two sessions for same job. We should keep the lock until session is created.

Reply:
We'd have to keep a write lock. Not sure if that's desirable when this particular code path should be read-heavy. That's why this check is here as you mentioned: https://github.com/livepeer/go-livepeer/pull/1124#discussion_r357898212

Reply:
OK, but then we can in `createSession`, after taking write lock check again if session exists, and return it?

Reply:
Fixed in https://github.com/livepeer/go-livepeer/pull/1124/commits/b9e09ffeac9470c420a0c489cb4df12f4bb2df8d